### ISSUE TITLE
Document Observability support for Knative components and Knative services

### ISSUE THREAD

**Issue by skonto** [knative/docs#2931]
**Describe the change you'd like to see**
In the [related WG meeting](https://docs.google.com/document/d/1Y7rug0XshcQPdKzptdWbQLQjcjgpFdLeEgP1nfkDAe4/) we discussed that users may require guidance about what metrics to use when monitoring Knative services. This has been raised before here #2070. The purpose of this feature is to capture the involved personas, what are the use cases and as a result what is the minimum info to expose in docs wrt observability not just metrics.
The personas can be splitted into roughly two groups in terms of reqs:
a) developer/tester/service delivery 
These people develop and deploy knative services and they would like to know what to monitor to make sure
everything is running smoothly in staging or development envs. They want to understand key Knative features like
Serving auto-scaling and measure performance before shipping their service.If they use Eventing they want to know how their processing graph behaves (sources/sinks/dispatchers/brokers etc) and understand if their deployed artifacts are performing well. But they also might want to understand networking performance/behavior when a service mesh is used etc. 
These people assume that Knative control plane and infrastructure is in place to run their services and debug them using logs, metrics and traces.

b) SREs/Operations/support engineers
These people are responsible for operating/debugging/automating Knative services and Knative components themselves in production. As a minimum they need support for the [4 golden signals ](https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#:~:text=The%20four%20golden%20signals%20of,system%2C%20focus%20on%20these%20four.) to set proper alarms etc. For example how can we help them detect if their setup is healthy.
They need to understand how Knative components and knative services integrate with the existing infrastructure at all levels. In that perspective, they need to know what reqs Knative imposes, if any, when integrating with existing infrastructure in terms of logging, metrics acquisition and tracing. 

Notes: 

1) Part of the observability functionality will be provided by vendors building products on top of Knative and thus they can also been seen as a separate persona.
2) It is important to help people (any persona) with their getting started UX.

The proposed minimum items to add to the docs are:

- What metrics are important for the different personas, a description should be available along with a sample example of what is considered a good value and what is not. What are the metrics needed to debug common issues like latency, saturation etc.
It would be nice to have a sample app that demonstrates all the above.
- What logs/traces the different personas need to gather to monitor the different Knative components.

Note: Any change to the above should be done with a followup change in docs as part of the PR work upstream.
I believe metrics are not expected to change that often.

/cc @abrennan89 @evankanderson @dprotaso

**Comment by upodroid** [comment]
I like the route that Istio took which is to provide the grafana dashboards and Service Monitor resources for end users to deploy them in to their prometheus installations. I'm happy to contribute them

**Comment by iancoffey** [comment]
I am game to help develop this solution as it is necessary to solve the [kfserving metrics situation](https://github.com/kubeflow/kfserving/pull/1162). 👍 😄

**Comment by skonto** [comment]
@iancoffey does Kfserving offer any tooling for setting up prometheus and/or grafana for the getting started UX at least? Did it use to rely on Knative for that setup?  Do you think this tooling could be part of a shared repo, if so how this could be maintained? The idea for describing stuff in docs is oriented more for providing high level info on what metrics tell you and not copying the old prometheus setup scripts.

**Comment by iancoffey** [comment]
> Did it use to rely on Knative for that setup?

kfserving docs relied on the observability metrics bits that was removed, and thats the genesis of my interest - trying to understand that gap better, and how to help solve it.

> Do you think this tooling could be part of a shared repo, if so how this could be maintained? 

Its an interesting idea, maybe a small experimental project. It would be great to have something to plug the gap until there are longer term options down the road. I wonder what kfserving folks would think of that.

**Comment by skonto** [comment]
>  I wonder what kfserving folks would think of that.
Pinged @evankanderson for that matter as well.

**Comment by iancoffey** [comment]
Ive assembled a simple yaml+config solution to get a knative grafana dashboard up and running here > https://github.com/iancoffey/knative-metrics. The intention is to unblock kfserving metrics for the time-being. I can add all the relevant dashboards if theres interest in developing this stopgap further.

**Comment by yuzisun** [comment]
> Ive assembled a simple yaml+config solution to get a knative grafana dashboard up and running here > https://github.com/iancoffey/knative-metrics. The intention is to unblock kfserving metrics for the time-being. I can add all the relevant dashboards if theres interest in developing this stopgap further.

I like the idea maybe this can temporarily live in KFServing as additional monitoring plugin users can install before we figure out a long term maintenance model with knative team.

**Comment by iancoffey** [comment]
OK I shall open this in a PR on kfserving repo next 👍

**Comment by skonto** [comment]
@iancoffey thanks for hosting this, I went through the yaml files, I found some incosistencies wrt removed scrape config, check [this](https://github.com/knative/serving/blob/d2ec5468b2fd5c018d314eba24941e3f22331264/config/monitoring/metrics/prometheus/100-prometheus-scrape-config.yaml#L112) vs [this](https://github.com/iancoffey/knative-metrics/blob/8e1015d211d3eabf55bc40fe5ca82f190625ef1d/manifests/20-prometheus.yaml#L367) line.
I will ping people on the monitoring channel maybe we can store this under knative.

**Comment by dprotaso** [comment]
Following up: @iancoffey switched teams and isn't involved with KFServing anymore.

>  I'm happy to contribute them

@upodroid are you interested in _maintaining_ these contributions across releases? We ideally would like someone to be regularly involved - meaning fixes issues, write some documentation on how to use these monitors. We just want to avoid these going stale again and being unmaintained.

/remove-lifecycle stale

**Comment by upodroid** [comment]
I'm up for it.

I have joined the google group. 

This fits the Operations WG nicely.

**Comment by skonto** [comment]
I can also help @upodroid :) We should discuss where to host such artifacts.

**Comment by upodroid** [comment]
great, i'll see you at the Operations WG today.

**Comment by dprotaso** [comment]
Following up from the operations WG meeting

We're going to create a sandbox repo and revive parts of the monitoring bundle:
https://github.com/knative/community/issues/654
https://github.com/knative/community/pull/653

After revival we'll drive improvements and potential publishing of dashboards to grafana.com
